src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+    return f(x, std::forward<Ts>(args)...);
+  }
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+};
+
+template <typename T, typename F>
+capture_impl<T, F> capture(T &&x, F &&f) {
+  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+}
+
+/// A helper for creating a snapshot view of the stats for a handler.
+/// This acquires a reader lock on the provided guarded handler stats, and creates a
+/// lockless copy of the stats.
+HandlerStats to_stats_view(const GuardedHandlerStats &stats) {
+  absl::ReaderMutexLock lock(&(stats.mutex));
+  return HandlerStats(stats.stats);
+}
+
+/// Set stats collection flag based on corresponding Ray config flag.
+const bool io_context_proxy::stats_collection_enabled_ =
+    RayConfig::instance().asio_event_loop_stats_collection_enabled();
+
+void io_context_proxy::post(std::function<void()> handler, const std::string &name) {
+  if (!stats_collection_enabled_) {
+    return boost::asio::io_context::post(std::move(handler));
+  }
+  {
+    absl::WriterMutexLock lock(&(global_stats_.mutex));
+    global_stats_.stats.cum_count++;
+    global_stats_.stats.curr_count++;
+  }
+  // Get this handler's stats.
+  mutex_.ReaderLock();
+  auto it = post_handler_stats_.find(name);
+  mutex_.ReaderUnlock();
+  if (it == post_handler_stats_.end()) {
+    // Lock the table until we have added the entry. We use try_emplace and handle a
+    // failed insertion in case the item was added before we acquire the writers lock;
+    // this allows the common path, in which the handler already exists in the hash table,
+    // to only require the readers lock.
+    absl::WriterMutexLock lock(&mutex_);
+    const auto pair =
+        post_handler_stats_.try_emplace(name, std::make_shared<GuardedHandlerStats>());
+    if (pair.second) {
+      it = pair.first;
+    } else {
+      it = post_handler_stats_.find(name);
+      // If try_emplace failed to insert the item, the item is guaranteed to exist in
+      // the table.
+      RAY_CHECK(it != post_handler_stats_.end());
+    }
+  }
+  {
+    absl::WriterMutexLock lock(&(it->second->mutex));
+    it->second->stats.cum_count++;
+    it->second->stats.curr_count++;
+  }
+  const auto start = std::chrono::steady_clock::now();
+  // References are only invalidated upon deletion of the corresponding item from the
+  // table, which we won't do until this io_context is deleted. Provided that
+  // GuardedHandlerStats synchronizes internal access, we can concurrently write to these
+  // stats from multiple threads without acquiring a table-level readers lock in the
+  // callback.
+  auto &stats = it->second;
+  // TODO(Clark): Use generalized lambda capture to move `handler` into the lambda once
+  // we move to C++14.
+  auto wrapped_handler = capture(
+      std::move(handler), [this, name, start, stats](std::function<void()> handler) {
```

Comment:
Although `name` and `start` are cheap to copy, we might want to move `stats` into the lambda as well.